### PR TITLE
add rancher-compose sub-generator

### LINKED ISSUE #5138: RancherOS support
Following the comments:
- https://github.com/jhipster/generator-jhipster/issues/5128#issuecomment-277940736
- https://github.com/jhipster/generator-jhipster/issues/5128#issuecomment-278147543
- https://github.com/jhipster/generator-jhipster/issues/4955#issuecomment-278143008

It would be cool to support http://rancher.com/rancher-os/ 
The documentation: https://docs.rancher.com/rancher/v1.4/en/

@PierreBesson suggested a new sub generator, but as I never use rancher, I let you guys take the lead on this.

### PR DESCRIPTION
**Pull request by stevehouel**
implement new rancher-compose sub-generator who will create docker-compose and rancher-compose specific files depending on their configuration. Il also provide a way to add the default rancher loadbalancer for front part (gateway or monolith app).

Fix #5138

### CODE REVIEW

**Comment by PierreBesson** [comment]
Thanks for the contribution @stevehouel !

We will have to test it with many configurations before merging. Also we should mark it as BETA.
I'm a bit worried about code duplication. Especially as we are already changing the docker-compose configuration for example by removing things like "container_name".

From the docs, I thought that Rancher only supported the docker-compose v1 syntax. But if it can support v2 and beyond maybe we could share part of the code with the docker subgen.

**Comment by stevehouel** [comment]
@PierreBesson You'r Welcome :)
For your information Rancher is based on libcompose lib (https://github.com/docker/libcompose) who is supporting docker-compose v2 syntax since April. It does not support v3 yet that why I did some duplicate code due to issue #5135, I did not want to create some dependencies between those two (docker-compose v3 have some really nice features as service restart etc). But if you want to merge both I can do it. 

Sorry for the BETA I just add it on the sayHello text. Could you tell me where to add it ? For my test I check on the last rancher version. I stay in touch when you will check those configs on your Rancher.

File: generators/rancher-compose/index.js [hunk 1]
```diff
@@ -0,0 +1,382 @@
+/*
+ * Copyright 2016 Google Inc. All Rights Reserved.
```
Comment (jdubois) [inline generators/rancher-compose/index.js]: why do you have a Google copyright here?

Comment (stevehouel) [inline generators/rancher-compose/index.js]: My bad. I have used k8n subgen as ref so I did not check it. I will remove it

**Comment by jdubois** [comment]
Yes this code duplication is annoying, but as they are not exactly the same things, it's going to be hard to avoid... @stevehouel can you check if you could have less duplication, and re-use the existing Docker/Kubernetes features? I'm not sure if that's possible, so don't hesitate to send feedback on this.

**Comment by stevehouel** [comment]
Ok @jdubois I will try to do some refactor with existing Docker-Compose feature. I did not use any code from k8n part. Just to be clear, do I keep in mind that docker-compose subgen and mine will have different lifecycle ?

**Comment by jdubois** [comment]
Yes you will have different lifecycles, then if you have a lot of common code it's really not good to copy/paste everything

**Comment by stevehouel** [comment]
Sure I already start to do that refacroring for all generators using docker
components and I prepare an other gen for mesos marathon stuff
Le ven. 17 févr. 2017 à 14:38, Julien Dubois <notifications@github.com> a
écrit :

**Comment by PierreBesson** [comment]
@stevehouel I have started to work on some changes that I want to do on your branch. I will do a PR on your fork next week. It will include the following changes :
- remove container names and external links
- remove prometheus monitoring as it doesn't work because it cannot load it's config
- fix consul

**Comment by stevehouel** [comment]
Ok @PierreBesson, I am waiting your PR to merge everything and update that one.

**Comment by PierreBesson** [comment]
I am still working on this. This is really interesting... I'm working on some sidekick containers to load the config as I need to load a config file for consul to work properly.

**Comment by stevehouel** [comment]
@PierreBesson Any news ? Need some help for what you are doing ?

**Comment by PierreBesson** [comment]
@stevehouel I did some progress but I was busy at work this week. I have found that there is still quite a bit of work to do on this subgen, especially for microservices support (setup registries with config, etc...). So there are definitely many things to polish here...

I will do my PR when I have the time. In the meantime if you want to do some work on this you can prepare the documentation page on the jhipster/jhipster.github.io repo. This is very important and the team doesn't have a lot of time to do it so I'm trusting you on this. Thank you.

**Comment by stevehouel** [comment]
Yes @PierreBesson I already start documentation on this subgen. If you need some helps for config I am here :) .

**Comment by cbornet** [comment]
Since Rancher has it's own service discovery, could it be used as an alternative to eureka/consul ? The interest for me is that it would allow to add non-JHipster/Java services easily.

**Comment by stevehouel** [comment]
@cbornet I will check it and try to bind JHipster to it. But I think this shoud be an other update to my PR. I keep you all in touch about that

**Comment by jdubois** [comment]
We need to have a service discovery that is compatible with Spring Cloud -> we use the same Spring Cloud API for Eureka and Consul. For example, this is needed for scaling Hazelcast.
So for me the JHipster Registry should be a better option, and we should have it on the Rancher store.

**Comment by PierreBesson** [comment]
@cbornet In any case if we support other service discovery systems it should be supported by the Spring Cloud abstraction so that we don't have to do too much work to integrate with it.

Kubernetes would be a good candidate as some people have already worked on this (see https://github.com/fabric8io/spring-cloud-kubernetes ). An alternative is to use no service discovery at all and rely on Ribbon serverList, (See [my stackoverflow post](http://stackoverflow.com/questions/40905303/integrate-eureka-with-kubernetes-issue/41000005#41000005)).

However I agree with @jdubois, we should really push people to use the Registry as it works everywhere (also lots of cool features are on their way... 😄 ).

@stevehouel I'm finishing my PR for the rancher-compose subgen as soon as I can. Have you made progress on the docs.

**Comment by stevehouel** [comment]
@PierreBesson Yes my PR is ready for the doc, just waiting your update to finalize it with configuration part.

**Comment by cbornet** [comment]
@PierreBesson @jdubois agreed that we should use only Spring-Cloud compatible service discoveries. Now the problem I have with JHipster Registry/Eureka/Ribbon is that you're bound to JVM languages or you need dedicated client code in other languages. At least Consul make it more technology transparent.

**Comment by deepu105** [comment]
Guys I'm sorry but this sub generator would need some updates to be compatible with the new lint rules and yeoman 1.x . can you plz rebase @stevehouel

**Comment by stevehouel** [comment]
Sure i will do it

**Comment by pascalgrimaud** [comment]
add 'on-hold' label, waiting proposals from @PierreBesson

**Comment by stevehouel** [comment]
My bad. Pull request was merged before I commit lint correction. I fix this and commit ASAP

**Comment by jdubois** [comment]
FYI I'm including a slide on this for my Devoxx France talk :-)

**Comment by stevehouel** [comment]
Great !!  Thx @jdubois

**Comment by deepu105** [comment]
@jdubois @stevehouel @PierreBesson why is this on hold?

**Comment by stevehouel** [comment]
We re finishing to test all case on rancher server. For now i just need to
finalize one test for ms architecture and cassandra db

**Comment by pascalgrimaud** [comment]
@stevehouel : don't put to much effort on Cassandra. In the Kubernetes sub generator (still in Beta), we didn't support Cassandra yet, and I'm not sure we'll do it in the future.

**Comment by stevehouel** [comment]
Ok so I think we can merge this PR. I let @PierreBesson validate and do it
Le dim. 2 avr. 2017 à 10:52, Pascal Grimaud <notifications@github.com> a
écrit :

**Comment by jdubois** [comment]
Yes, don't do Cassandra: the only real use case for Cassandra is if you have 10+ nodes, and in that case it requires ops procedures which are far beyond what we should do.

**Comment by stevehouel** [comment]
What i have done is based on docker-compose module so cassandra is already
done i juste need to test and validate.
Le dim. 2 avr. 2017 à 11:54, Julien Dubois <notifications@github.com> a
écrit :

**Comment by PierreBesson** [comment]
@stevehouel, no Cassandra is not going to work with Rancher out of the box since it relies on a local volume to launch CQL scripts so you would need a special sidekick to do it. I had thought of doing it for k8s then I realised that it's not worth it.

I will do some testing next week and then merge your PR before wednesday when we will do a release.

**Comment by stevehouel** [comment]
Ok great @PierreBesson.

Le dim. 2 avr. 2017 à 17:45, Pierre Besson <notifications@github.com> a
écrit :

**Comment by PierreBesson** [comment]
Nice work @stevehouel !

**Comment by stevehouel** [comment]
Great !! I push documentation updated asap
Le mar. 4 avr. 2017 à 18:30, Pierre Besson <notifications@github.com> a
écrit :

**Comment by stevehouel** [comment]
Your welcome @PierreBesson. Mesos marathon module is coming soon 😋
Le mar. 4 avr. 2017 à 18:32, Steve HOUEL <steve.houel.perso@gmail.com> a
écrit :

**Comment by jdubois** [comment]
@stevehouel you now need to add the documentation, next to https://jhipster.github.io/kubernetes/ - just copy/paste the file and do your documentation, don't worry too much about the menu, I'll do it.

Remember the Zen koan: If a tree falls in a forest and no one is around to hear it, does it make a sound?

**Comment by stevehouel** [comment]
Documentation was already done :) i was just waiting PR merge and what
@PierreBesson did on configuration to finish it
Le mar. 4 avr. 2017 à 18:41, Julien Dubois <notifications@github.com> a
écrit :

**Comment by jdubois** [comment]
OK then @stevehouel just do the PR on the website, I'll review & and merge :-)

**Pull request merged**